operator_rolling_upgrades.md

```diff
@@ -0,0 +1,54 @@
+# Operator Rolling Upgrades
+Optionally allow a Infinispan CR cluster to be upgraded with zero downtime and no loss of data.
+
+## Spec
+Add a field to the CR spec that determines whether an upgrade should be `rolling` or `gracefulShutdown`. The latter is the current approach that does not allow data migration.
+
+```yaml
+apiVersion: infinispan.org/v1
+kind: Infinispan
+metadata:
+  name: example-infinispan
+spec:
+  upgrades: rolling
+```
+
+## Procedure
+When an upgrade is possible, the operator performs the following for each Infinispan CR:
+
+1. Create a new StatefulSet for the target cluster based upon the CR spec, wait until the cluster is formed
+
+2. Retrieve all cache configurations from the cluster in the original Statefulset
+
+3. Append a remote-store configuration to each of the ^ configurations and create cache target cluster:
+
+```xml
+<remote-store xmlns="urn:infinispan:config:store:remote:12.0"
+                 cache="<cache-name>" 
+                 protocol-version="<protocol-version>"> 
+      <remote-server host="<ip-of-pod>" port="11222"/> 
+   </remote-store>
+```
+
+4. Switch clients to the target cluster.
+
+5. Update the CR spec.StatefulSetName to new Statefulset
+
+6. Sync data from source for all caches. Execute the following on the target cluster
+
+```
+POST /v2/caches/{cacheName}?action=sync-data
+```
+
+7. Disconnect remote-store on the target cluster for all caches
+```
+POST /v2/caches/{cacheName}?action=disconnect-source
+```
+
+8. Delete the source StatefulSet
+
+## Switching Clients
+- How do we ensure that the single-port service only points to the target cluster?
```

Comment:
Service targets pods created by `StatefulSet`, you'll change the selector of the service to match only pods of the new `StatefulSet`.

Eg.: Old StatefulSet would be named `infinispan-11.0.1` and the new one `infinispan-12.0.0`. That would be reflected in the pod labels. Thus making it simple to correctly reference by new ping service and update of the `single-port` service

Reply:
Is it possible to update the selector of the service with zero downtime? I'm just wondering about HR client failover.

Reply:
I believe you can just switch the labels once the target cluster is ready. Once you establish a TCP connection from a client to a server, it's point-to-point. The service is only required when you initiate such a connection. 

This means that the switching the clients from the old to the new will happen automatically as soon as you start wiping the old cluster out.

Reply:
Just thinking on it a bit more - maybe you should switch the labels as soon as you start the upgrade procedure (not at the end of it)?

Reply:
We need to wait for the cache's to be defined on the target cluster with the remote-store configuration, we can then switch the labels before proceeding with the sync. Without the cache config on the target cluster, the caches won't exist and it won't be possible to retrieve old values from the source cluster.

Reply:
> Is it possible to update the selector of the service without unavailability? I'm just wondering about HR client failover.

It should take effect immediately or at least there shouldn't be a gap when the service isn't connected to either cluster. But I've never tested it.